File: leetcode/056.merge-intervals.py
#!/usr/bin/env python3
'''
leetcode #056 merge-intervals 合并区间
https://leetcode-cn.com/problems/merge-intervals/description/
'''

# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        if len(intervals) == 0:
            return []
        intervals.sort(key = lambda a: a.start)
        ret = []
        pre = intervals[0]
        for i in range(1, len(intervals)):
            if intervals[i].start <= pre.end:
                pre.end = max(pre.end, intervals[i].end)
            else:
                ret.append(pre)
                pre = intervals[i]
        ret.append(pre)
        return ret

File: leetcode/056.merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function (intervals) {
  return intervals
    .sort((a, b) => a[0] - b[0])
    .reduce((acc, cur) => {
      if (!acc.length) {
        acc.push(cur)
        return acc
      }

      const last = acc[acc.length - 1]
      if (last[1] >= cur[0]) {
        last[0] = Math.min(last[0], cur[0])
        last[1] = Math.max(last[1], cur[1])
        return acc
      }
      acc.push(cur)
      return acc
    }, [])
}
